### ISSUE TITLE
`Style/TrailingCommaIn{Array,Hash}Literal` should not put a trailing comma immediately preceding a `}`/`]`

### ISSUE THREAD

**Issue by sambostock** [rubocop/rubocop#9935]
## Context

As far as I am aware, the rationale for trailing commas has always been:
- they make for cleaner diffs when adding/removing elements
  ```diff
   {
     foo: 123,
  +  bar: 456,
   }
  ```
  vs
  ```diff
   {
  -  foo: 123
  +  foo: 123,
  +  bar: 456
   }
  ```
- they avoid the risk of accidentally forgetting to add a comma following the proceeding element when appending a new one
  ```diff
   {
     foo: 123 # ❌
  +  bar: 456
   }
   ```

Therefore, a trailing comma proceeding a `}` on the same line serves no purpose, as the line would still need to be edited to change values, trailing comma or not:

```diff
 { foo: 12, bar: 34,
-  biz: 56, baz: 78, }
+  biz: 56, baz: 78,
+  more: 90, }
```

```diff
 { foo: 12, bar: 34,
-  biz: 56, baz: 78 }
+  biz: 56, baz: 78,
+  more: 90 }
```

## Expected behavior

`Style/TrailingCommaInHashLiteral`, when using `consistent_comma`, should consider the following code correct:

```ruby
{ foo: 12, bar: 34,
  biz: 56, baz: 78 }
```

## Actual behavior

`Style/TrailingCommaInHashLiteral`, when using `consistent_comma`, requires the following change

```diff
 { foo: 12, bar: 34,
-  biz: 56, baz: 78 }
+  biz: 56, baz: 78, }
```

<details><summary><code>rubocop --debug</code> output</summary>

```
$ bundle exec rubocop --only Style/TrailingCommaInHashLiteral --debug
For ~/src/local/rubocop-bug: configuration from ~/src/local/rubocop-bug/.rubocop.yml
Default configuration from ~/.gem/ruby/2.7.1/gems/rubocop-1.18.3/config/default.yml
[...]
Inspecting 2 files
Scanning ~src/local/rubocop-bug/Gemfile
Loading cache from ~/.cache/rubocop_cache/8909d50e923b0c43f2209673b32e44a0bb04a8f9/1acc38264efb39768ff7b6c492dd2cbf87baf7ae/28c790a0a9c5b27103db82cc6c4bc236a59e3112
.Scanning ~/src/local/rubocop-bug/script.rb
Loading cache from ~/.cache/rubocop_cache/8909d50e923b0c43f2209673b32e44a0bb04a8f9/1acc38264efb39768ff7b6c492dd2cbf87baf7ae/005dc2221bf5756dc9e26d1f4166040e37b98ff1
C

Offenses:

script.rb:2:12: C: [Correctable] Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.
  uiz: 56, baz: 78 }
           ^^^^^^^

2 files inspected, 1 offense detected, 1 offense auto-correctable
Finished in 0.4630139999790117 seconds
```
</details>

**This same issue is present in `Style/TrailingCommaInArrayLiteral`.**

## Steps to reproduce the problem

- ```yml
  # .rubocop.yml
  Style/TrailingCommaInHashLiteral:
    EnforcedStyleForMultiline: consistent_comma
  ```
- ```ruby
  # example.rb
  { foo: 12, bar: 34,
    biz: 56, baz: 78 }
  ```

## RuboCop version

```sh
$ bundle exec rubocop -V
1.18.3 (using Parser 3.0.2.0, rubocop-ast 1.8.0, running on ruby 2.7.1 x86_64-darwin19)
```

**Comment by koic** [comment]
I think @sambostock's opinion makes sense, but it will change the current behavior.
https://github.com/rubocop/rubocop/blob/v1.18.3/spec/rubocop/cop/style/trailing_comma_in_hash_literal_spec.rb#L180-L197

@rubocop/rubocop-core Are there any concerns about changing this behavior?
For example, the current behavior is consistent with the naming `EnforcedStyleForMultiline: consistent_comma`. Therefore some users may be expecting the current behavior.

**Comment by sambostock** [comment]
Just had this pop up on another PR and wanted to give some more examples of corrections that I don't think are appropriate:

```diff
 something = [{
   # ...
 }, {
   # ...
-}]
+},]

 {
   something: 123,
   attributes: { identifier: "abc",
                 info: { foo: { id: "123" }, bar: { id: "123" },
-                        biz: { baz: "xyz" } } },
+                        biz: { baz: "xyz" }, }, },
   blah: 'blah',
 }

 {
   "fields" => [
     { "id" => "1415",
-      "value" => "example.com" },
+      "value" => "example.com", },
   ],
 }

 report = {
   "entries" => [{
     "Foo" => "abc",
   }, {
     "Bar" => "def",
-  }],
+  },],
 }
```

@koic, do you think this requires the addition of a new option, or that we should change `consistent_comma`'s behaviour?

**Comment by sambostock** [comment]
**I just realized this also applies to `Style/TrailingCommaInArrayLiteral`.** In fact, one of my examples above demonstrates it.

IMO, this is a **bug** in both rules because the commas are not at the end of a line (ignoring comments) or delimiting between entries.

**Comment by dmitrytrager** [comment]
At the same time for 
```
Style/TrailingCommaInArguments:
  EnforcedStyleForMultiline: consistent_comma
```
this syntax should raise a warning:
```
create( :entity,
      name: 'name',
      key: 'key',)
```
And it does not, only `comma` style does

Rubocop version 1.31.2

**Comment by pointlessone** [comment]
A related issue with `Style/TrailingCommaInArrayLiteral` with `EnforcedStyleForMultiline: comma` complains about this:

```ruby
[
  1, 2, 3,
  4, 5, 6,
]
```

> Avoid comma after the last item of an array, unless each item is on its own line.

What I ultimately want to achieve is minimal diffs. I want commas at the end of line in multiline… everything. And I'm fine with no coma if closing punctuation is at the end of line after elements but that is not required as a different rule can handle that.

The current solution is not ideal though. I'm fine with putting a couple of elements on the same line. The rule though insists on a single element per line. Maybe a new rule can deal with the number of elements per line and `TrailingComma*` rules only pay attention to commas?

**Comment by akimd** [comment]
This issue raises very good points, and several of us hit this limitation.

In my case it was so much of a problem that we entirely disabled these cops.  Which is a pity.

It would be nice if this issue could get some attention.  Thanks!

**Comment by flavorjones** [comment]
I've proposed a new style "diff_comma" in https://github.com/rubocop/rubocop/pull/13806.
